src/python-bindings/htcondor2/job_event_log.cpp

```diff
@@ -1 +1,219 @@
+#endif
+
+
+class JobEventLogGlobalLockInitializer {
+    public:
+        JobEventLogGlobalLockInitializer() {
+#if defined(WIN32)
+            MODULE_LOCK_MUTEX_INITIALIZE( & mutex );
+#else
+            mutex = MODULE_LOCK_MUTEX_INITIALIZER;
+#endif
+        }
+
+        MODULE_LOCK_MUTEX_TYPE mutex;
+} jobEventLogGlobalLock;
+
+
+static PyObject *
+_job_event_log_next( PyObject *, PyObject * args ) {
+    PyObject * self = NULL;
+    PyObject_Handle * handle = NULL;
+    long deadline = 0;
+
+    if(! PyArg_ParseTuple( args, "OOl", & self, (PyObject **)& handle, & deadline )) {
+        // PyArg_ParseTuple() has already set an exception for us.
+        return NULL;
+    }
+
+    auto wful = (WaitForUserLog *)handle->t;
+
+
+    ULogEvent * event = NULL;
+    // Must not be declared inside the invisible scope in which we allow
+    // other Python threads to run.
+    ULogEventOutcome outcome;
+
+    Py_BEGIN_ALLOW_THREADS
+    MODULE_LOCK_MUTEX_LOCK( & jobEventLogGlobalLock.mutex );
+
+        if( deadline ) {
+            time_t now = time(NULL);
+
+            if( deadline <= now ) {
+                // If the deadline is now, or has passed, poll for an event.
+                outcome = wful->readEvent( event, 0, false );
+            } else {
+                // If the deadline has yet to pass, block until it does.
+                outcome = wful->readEvent( event, (deadline - now) * 1000, true );
+            }
+        } else {
+            // If there isn't a deadline, block forever.
+            outcome = wful->readEvent( event, -1, true );
+        }
+
+    MODULE_LOCK_MUTEX_UNLOCK( & jobEventLogGlobalLock.mutex );
+    Py_END_ALLOW_THREADS
+
+    switch( outcome ) {
+        case ULOG_OK: {
+            // Obtain the serialized form of the event.
+            std::string event_text;
+
+            int fo = USERLOG_FORMAT_DEFAULT;
+            auto format = param( "DEFAULT_USERLOG_FORMAT_OPTIONS" );
+            if( format != NULL ) {
+                fo = ULogEvent::parse_opts( format, USERLOG_FORMAT_DEFAULT );
+                free( format );
+                format = NULL;
+            }
+
+            if(! event->formatEvent( event_text, fo )) {
+                PyErr_SetString( PyExc_RuntimeError, "Failed to convert event to string" );
+                return NULL;
+            }
+
+
+            // Obtain the event data.
+            ClassAd * eventAd = event->toClassAd(false);
+            if( eventAd == NULL ) {
+                // This was HTCondorInternalError in version 1.
+                PyErr_SetString( PyExc_RuntimeError, "Failed to convert event to ClassAd" );
+                return NULL;
+            }
+
+            // FIXME: Is this necessary, or does the Python side just need
+            // to use key other than `type`?
+            eventAd->Assign("type", event->eventNumber);
+            // FIXME: ditto
+            eventAd->Assign("timestamp", event->GetEventclock());
```

Comment:
The C++ ClassAd stores just an ISO8601 string (either local TZ or UTC) in attribute EventTime. The python code could fetch that and convert to a timestamp.

Reply:
We discussed this in person and decided that for now we'd go with maintaining backwards-compatibility with version 1, so that the `timestamp` property will just be a int.  There's an easy work-around (ask for the `EventTime` directly), and we could formally support that as some other `@property` function, e.g., `datetime`.